### PR TITLE
[LAN-7166] Merge fast mode changes from avi-benchmark to landing-ai/mmdet

### PR DESCRIPTION
**Pull request by shankar-landing-ai**
## Motivation

Merge fast mode changes from avi-benchmark to landing-ai/mmdet

## Modification

- Remove Custom Hooks from default_runtime.py and have it in each configuration file

    - `mmdetection/configs/_base_/default_runtime.py`
    - As we develop more custom hooks, some of them might not be need in all config files.
      It makes it easier to track them if they are in individual config files

- Modify Model trainer code in mmdet api to inject auto tuning in trainer

    - `mmdetection/mmdet/apis/train.py`
    - Created FastModeCheckpointHook to checkpoint best weights that give balanced performance
      in both train and dev set
    - Added autotuning of samples_per_gpu (batch size) based on number of images in train set
    - Added default configuration values for pinned memory and persistent workers
    - Autotuned warmup epochs based on dataset size
    - Changed log interval to iters/epoch for averaging of losses
    - Changed EvalHook priority to BELOW_NORMAL to work in sync with FastModeCheckpoint
    - `mmdetection/mmdet/apis/__init__.py`
    - Add FastModeCheckpointHook to init.py

- Add top_f1 as an evaluation metric

    - `mmdetection/mmdet/core/avi.py`
    - top_f1 metric is the F&E metric used in front end to check performance of models
    - Also calculates instance_mAP at no or little extra cost

- Add pin_memory to dataloader builder

    - `mmdetection/mmdet/datasets/builder.py`
    - Leaving pin_memory attribute to False as recommended by PyTorch docs

- Update CustomDataset and AVICustomDataset to use avi metrics (instancemAP, top_f1)

    - `mmdetection/mmdet/datasets/custom.py`
    - `mmdetection/mmdet/datasets/avi_custom.py`
    - Add elif condition to check for top_f1 and add results to eval+results

- Update the base class of RepPointsHead to provide loss for validation runs

    - `mmdetection/mmdet/models/dense_heads/base_dense_head.py`
    - This is a possible bug in mmdet, training flag needs to true for val_loss

- Added code to handle evaluation on train set for calculating train dataset metrics

    - `/Users/shankar.local/Repos/mmdetection/tools/test.py`
    - We will change the pipeline for train set to test_pipeline so that there is no
      augmentation applied during eval

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is not covered by complete unit tests. Since we are potentially discontinuing this fork of mmdet, unit test cases will be handled in avi-repos 
3. However, end to end was conducted by consuming the repo as a submodule under avi-benchmark and train/eval results were validated for latency and performance

### CODE REVIEW

File: mmdet/apis/train.py [hunk 1]
```diff
@@ -175,28 +216,30 @@ def train_detector(model,
...
+        )
+        eval_cfg = cfg.get("evaluation", {})
+        eval_cfg["by_epoch"] = cfg.runner["type"] != "IterBasedRunner"
         eval_hook = DistEvalHook if distributed else EvalHook
         # In this PR (https://github.com/open-mmlab/mmcv/pull/1193), the
         # priority of IterTimerHook has been modified from 'NORMAL' to 'LOW'.
         runner.register_hook(
-            eval_hook(val_dataloader, **eval_cfg), priority='LOW')
+            eval_hook(val_dataloader, **eval_cfg),
+            priority="BELOW_NORMAL",
```
Comment (jckhang) [inline mmdet/apis/train.py:238]: Can you add a note why you changed from LOW to BELOW_NORMAL

File: mmdet/apis/train.py [hunk 2]
```diff
@@ -206,3 +249,194 @@ def train_detector(model,
...
+        eval_config: dict = None,
+        train_margin: float = 0.03,
+        val_margin: float = 0.05,
+        extra_epochs: int = 2,
+        keep_best_only: bool = False,
+    ) -> None:
+        """Custom Checkpinting Hook for balancing speed and performance in fast mode.
+            The best checkpoint will be saved as best.pth
+        Args:
+            runner (obj:`EpochBasedRunner`): Epoch based Runner.
```
Comment (jckhang) [inline mmdet/apis/train.py:267]: runner is not an argument

File: mmdet/apis/train.py [hunk 3]
```diff
@@ -206,3 +249,194 @@ def train_detector(model,
...
+            best_val_epoch = np.argmax(self.val_history)
+            if best_train_epoch == best_val_epoch:
+                return adjusted_epoch(best_val_epoch)
+
+            best_train_loss = min(self.loss_history)
+            self.best_val_metric = max(self.val_history)
+            # If the loss value at best val epoch is within the limits of best loss,
+            # return best val_epoch
+            loss_at_best_val_epoch = self.loss_history[best_val_epoch]
+            loss_deviation = round(abs(loss_at_best_val_epoch - best_train_loss), 5)
```
Comment (jckhang) [inline mmdet/apis/train.py:372]: doesn't need abs... loss_at_best_val_epoch >= best_train_loss

File: mmdet/apis/train.py [hunk 4]
```diff
@@ -206,3 +249,194 @@ def train_detector(model,
...
+                Default: True
+
+        """
+
+        self.eval_keys = ["mAP", "recall", "instance_mAP", "top_f1"]
+        self.loss_key = "loss"
+        if eval_config is None:
+            eval_config = {"start": 9, "metric": ["instance_mAP"]}
+        self.start_eval_epoch = eval_config["start"]
+        if not isinstance(eval_config["metric"], str):
```
Comment (jckhang) [inline mmdet/apis/train.py:288]: what if metric is None, do we use a default key?

Comment (jckhang) [reply mmdet/apis/train.py:288]: metric = None, 0 will fail  this line

File: mmdet/apis/train.py [hunk 5]
```diff
@@ -206,3 +249,194 @@ def train_detector(model,
...
+
+        # Find Best epoch with extended run
+        # If extended run yielded better validation metric return corresponding epoch
+        best_val_metric_extended = max(self.val_history[-self.extra_epochs :])
+        if best_val_metric_extended >= self.best_val_metric:
+            return adjusted_epoch(np.argmax(self.val_history[-self.extra_epochs :]))
+
+        # Check if the top extended val metric is within limits of best regular run
+        # top validation metric if not return epoch with best loss from extended run
+        val_deviation = round(abs(self.best_val_metric - best_val_metric_extended), 5)
```
Comment (jckhang) [inline mmdet/apis/train.py:386]: Curious, why in regular flow, we check train deviation but in extended runs, we are checking validation deviation?

File: mmdet/apis/train.py [hunk 6]
```diff
@@ -206,3 +249,194 @@ def train_detector(model,
...
+            for i, ckpt_file in enumerate(purge_file_list):
+                if (
+                    file_client.isfile(ckpt_file)
+                    and (i + self.start_eval_epoch) != best_epoch
+                ):
+                    file_client.remove(ckpt_file)
+
+    @master_only
+    def return_best_checkpoint_epoch(self) -> int:
+        def adjusted_epoch(best_epoch: int, val_min: bool = True) -> int:
```
Comment (jckhang) [inline mmdet/apis/train.py:348]: Curious, what is val_min? train_history and validation_history has 1 off? 
can we use int(val_min)

File: mmdet/apis/train.py [hunk 7]
```diff
@@ -206,3 +249,194 @@ def train_detector(model,
     elif cfg.load_from:
         runner.load_checkpoint(cfg.load_from)
     runner.run(data_loaders, cfg.workflow)
+
+
+@HOOKS.register_module()
+class FastModeCheckpointHook(Hook):
```
Comment (jckhang) [inline mmdet/apis/train.py:255]: This is very special and complex class. Can you add a TODO to add test cases?

File: mmdet/apis/train.py [hunk 8]
```diff
@@ -206,3 +249,194 @@ def train_detector(model,
...
+        # If extended run yielded better validation metric return corresponding epoch
+        best_val_metric_extended = max(self.val_history[-self.extra_epochs :])
+        if best_val_metric_extended >= self.best_val_metric:
+            return adjusted_epoch(np.argmax(self.val_history[-self.extra_epochs :]))
+
+        # Check if the top extended val metric is within limits of best regular run
+        # top validation metric if not return epoch with best loss from extended run
+        val_deviation = round(abs(self.best_val_metric - best_val_metric_extended), 5)
+        if val_deviation <= self.val_tolerance:
+            return adjusted_epoch(np.argmax(self.val_history[-self.extra_epochs :]))
```
Comment (jckhang) [inline mmdet/apis/train.py:388]: Are we argmaxing on only extended runs? this looks like the last n epochs which for the first extended run, will go back to normal runs?

File: mmdet/apis/train.py (hunk 1 above)
Comment (shankar-landing-ai) [reply mmdet/apis/train.py:238]: Sure, it was done such that this callback runs between EvalHook and TextLoggerHook. We need to run it after EvalHook to make sure we have mAP values in log buffer and we can't run it after TextLoggerHook because log buffer gets cleared. I will add this info in the next commit

File: mmdet/apis/train.py (hunk 2 above)
Comment (shankar-landing-ai) [reply mmdet/apis/train.py:267]: I had it initially and then removed it. Will update the docstring, thanks for catching this

File: mmdet/apis/train.py (hunk 3 above)
Comment (shankar-landing-ai) [reply mmdet/apis/train.py:372]: Actually no, the loss_at_best_eval_epoch can be less than best loss and it has to be within the tolerance limit. Which is basically translates into "As long as loss is not significantly different it doesn't affect train mAP, so even if its less than best loss its ok to take that epoch"

Example 1 : best_loss = 0.0045, train_loss_at_best_eval_epoch = 0.0047 -> In this case its ok to take train_loss_at_best_eval_epoch as the difference is marginal (inside 0.03 tolerance).

Example 2: best_loss = 0.0045, train_loss_at_best_eval_epoch = 0.05 -> In this case its not ok to take train_loss_at_best_eval_epoch as the difference is significant (outside 0.03 tolerance).

File: mmdet/apis/train.py (hunk 4 above)
Comment (shankar-landing-ai) [reply mmdet/apis/train.py:288]: Actually metric cannot be None. It comes from config, if the config is missing then my line at 285 will handle it. If the eval config is missing "metric" param then the normal mAP will be calculated as seen here - https://github.com/landing-ai/mmdetection/blob/2a32508fd7a161bd74c79b0e0af5e5f75b8b4dfc/mmdet/datasets/avi_custom.py#L329

File: mmdet/apis/train.py (hunk 5 above)
Comment (shankar-landing-ai) [reply mmdet/apis/train.py:386]: Good question. Here is the reason. During the regular runs the criteria is, given the best mAP lets find the best train loss (alias for train mAP). If this is not attainable, thats when we trigger extended runs.

Once the extended runs are triggered based on my experiment results, the train loss improves and guaranteed to be lower than earlier runs. its like any neural network training the more epochs we run the train loss becomes lesser. However what happens during extended runs is we run into overfit mode so the criteria changes now. Given best train loss figure out best val mAP. Hope this answers your question

File: mmdet/apis/train.py (hunk 6 above)
Comment (shankar-landing-ai) [reply mmdet/apis/train.py:348]: I will write a doc string for this function and change the variable name.  I see now its confusing. Here is the explanation - Since we only calculate best epoch in subset of total epoch runs when we do argmax we need to adjust to index 0 (first epoch). This is the crux of this function. However for extended runs, we need to perform one adjustment as we pick only from extended run which is another subset. The only time we pick best train loss is on the extended run and we also consider the last epoch of regular run (epoch = 12). Since we open up the subset range for train loss, this function needs to know whether its adjusting for best val_epoch or best_train_epoch. By default we look for best "validation metric" minimum which is mentioned as val_min. When we look for best train loss we set this to False

**Review by jckhang — APPROVED** [review]
LGTM

File: maintanence.md [hunk 9]
```diff
@@ -37,3 +37,73 @@ The purpose of this repo is to provide important addons that are required for La
     - Use instance_mAP as evaluation metrics.
 
     - Load files from AWS S3.
+
+- Remove Custom Hooks from default_runtime.py and have it in each configuration file
+
+    - `mmdetection/configs/_base_/default_runtime.py`
```
Comment (jckhang) [inline maintanence.md:43]: avi is not reading from this file. the content of default_runtime can be found in https://github.com/landing-ai/avi/blob/master/avi/configs/templates/od_fast_less_accurate.yaml#L233-L236 and https://github.com/landing-ai/avi/blob/master/avi/configs/templates/od_slow_more_accurate.yaml#L329-L332

File: maintanence.md [hunk 10]
```diff
@@ -37,3 +37,73 @@ The purpose of this repo is to provide important addons that are required for La
...
+    - Autotuned warmup epochs based on dataset size
+    - Changed log interval to iters/epoch for averaging of losses
+    - Changed EvalHook priority to BELOW_NORMAL to work in sync with FastModeCheckpoint
+    - `mmdetection/mmdet/apis/__init__.py`
+    - Add FastModeCheckpointHook to init.py
+
+- Add top_f1 as an evaluation metric
+
+    - `mmdetection/mmdet/core/avi.py`
+    - top_f1 metric is the F&E metric used in front end to check performance of models
```
Comment (jckhang) [inline maintanence.md:63]: top_f1 metric doesn't need to come from us / mmdet library. avi-common only needs `prediction` of bbox, bbox_scores, bbox_labels for [this function](https://github.com/landing-ai/avi-common/blob/fd681a717873189d98d39ef9cd5f78cab54a840b/app/newflow/reporter/predict_and_report.py#L136) and avi-common computes f1

File: maintanence.md [hunk 11]
```diff
@@ -37,3 +37,73 @@ The purpose of this repo is to provide important addons that are required for La
...
+
+- Update the base class of RepPointsHead to provide loss for validation runs
+
+    - `mmdetection/mmdet/models/dense_heads/base_dense_head.py`
+    - This is a possible bug in mmdet, training flag needs to true for val_loss
+
+- Added code to handle evaluation on train set for calculating train dataset metrics
+
+    - `/Users/shankar.local/Repos/mmdetection/tools/test.py`
+    - We will change the pipeline for train set to test_pipeline so that there is no
```
Comment (jckhang) [inline maintanence.md:85]: I have something similar in [this PR](https://github.com/landing-ai/avi/pull/728/files#diff-74a1ac1629c0ac612bba00f246630c20c9b52994effe3a9303913886643d6d32R69). If it covers this topic, you can skip this task

**Review by Ivan-Zhou — APPROVED** [review]
I've validated this change end-to-end on my mislabel detection app
